### PR TITLE
Add Indeterminate State to Checkbox Kit

### PR DESCRIPTION
**Pull request by coleerikson**
#### Screens

indeterminate-doc

#### Breaking Changes

Most likely yes, still investigating. We're changing the `_checkbox.jsx` kit to a controlled component that uses state to manage checked and unchecked states.

#### Runway Ticket URL

https://nitro.powerhrg.com/runway/backlog_items/NUXE-229

#### How to test this

Locations in Nitro:

Rails (20 locations, 12 files):

/contact_center/app/views/contact_center/agent_stats/_qualifiers.html.erb
/email/app/views/email/email_campaigns/search_forms/_employee.html.erb
/nitro_component_transition/app/views/bootstrap/accessory_items/index/_table.html.erb
/nitro_component_transition/app/views/bootstrap/accessory_types/_form.html.erb
/nitro_component_transition/app/views/bootstrap/project_tasks/_install_queue_items_table.html.erb
/nitro_survey/app/views/nitro_survey/blasts/_prompt.html.erb
/recruiting/app/views/recruiting/candidate_applications/index.html.erb
/recruiting/app/views/recruiting/retention_report/index.html.erb
/sales/app/views/sales/reps/qualifiers.html.erb
/sales/app/views/sales/test_pages/mentor_group_bonuses.html.erb
/user_profile/app/views/user_profile/user/products.html.erb
/user_profile/app/views/user_profile/user/sources.html.erb
React (21 locations, 17 files):

/accounting/app/components/invoices/ListTable.js
/employee_health_check/app/components/HealthCheck/CheckboxInput.jsx
/employee_health_check/app/components/HealthCheckOptions/OptionsForm.jsx
/human_resources/app/components/HistoryFlyout/FlagNotes.jsx
/inventory/app/components/InventoryLocationForm/ProductSelection.js
/inventory/app/components/ReorderRuleListing/Row.js
/inventory/app/components/Search/CheckFilter.js
/inventory/app/containers/InventoryLocationForm.js
/inventory/app/containers/GenerateBarcodeForm/container.js
/nitro_react/app/components/CheckboxArrayFormField.jsx
/nitro_react/app/components/CheckboxFormField.jsx
/nitro_survey/app/components/SurveyTaker/WrappedInputs.jsx
/recruiting/app/components/CheckboxInput.jsx
/support_tickets/app/plugins/project/ProjectFormFields/ProductCheckbox.jsx
/support_tickets/app/plugins/project/Service/ServiceForm.js
/user_profile/app/components/MyProfile/CheckboxInput.jsx
/warranty/app/components/WarrantyServiceOrder/ProjectItemFields.jsx

#### Checklist:

- [x] **LABELS** Add a label: `enhancement`, `bug`, `improvement`, `new kit`, `deprecated`, or `breaking`. See [Changelog & Labels](https://github.com/powerhome/playbook/wiki/Changelog-&-Labels) for details.
- [x] **URGENCY** Please select a release milestone
- [x] **DEPLOY** Please add the `Milano` label when you are ready for a review.
- [x] **SCREENSHOT** Please add a screen shot or two.
- [x] **SPECS** Please cover your changes with specs.
- [x] **READ DOCS** Please make sure you have read and understand the [Playbook Release Process](https://github.com/powerhome/playbook/wiki/Playbook-Releases)

*The normal release cut off deadline is 3p EDT each week. Please reach out to the release team if you have an urgent request or need a release off cycle.*

### CODE REVIEW

File: playbook/app/pb_kits/playbook/pb_checkbox/_checkbox.jsx [hunk 1]
```diff
@@ -41,11 +43,22 @@ const Checkbox = (props: CheckboxProps) => {
...
 
+  const [ isChecked, toggleCheck ] = useState(checked)
+
+  useEffect(() => {
+    toggleCheck(checked)
+  }, [checked])
+
+  const handleClick = () => {
+    toggleCheck(!isChecked)
+    onChange()
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_checkbox/_checkbox.jsx:59]:
```suggestion
    onChange(isChecked)
```

**Review by thestephenmarshall — APPROVED** [review]

File: playbook/app/pb_kits/playbook/pb_checkbox/docs/_checkbox_indeterminate.html.erb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          classname: "check-js"
+        }) %>
+      </td>
+    </tr>
+  </tbody>
+<% end %>
+
+
+<script>
+  const indeterminateCheckbox = document.querySelector("#test-indeterminate-js");
```
Comment (xjunior) [inline playbook/app/pb_kits/playbook/pb_checkbox/docs/_checkbox_indeterminate.html.erb:43]: This kit seems like a good candidate to reuse the react kit

Comment (thestephenmarshall) [reply playbook/app/pb_kits/playbook/pb_checkbox/docs/_checkbox_indeterminate.html.erb:43]: I agree. We will have to get the form situation worked out to ensure that anyone using this kit will receive the same markup as expected by Rails within at least the `form_for` implementation. Some work is necessary to make that happen within the form kit itself.

**Review by jasoncypret — CHANGES_REQUESTED** [review]
Needs Rebased

**Comment by coleerikson** [comment]
@jasoncypret Locations in Nitro:

**Rails (20 locations, 12 files):**
- /contact_center/app/views/contact_center/agent_stats/_qualifiers.html.erb
- /email/app/views/email/email_campaigns/search_forms/_employee.html.erb
- /nitro_component_transition/app/views/bootstrap/accessory_items/index/_table.html.erb
- /nitro_component_transition/app/views/bootstrap/accessory_types/_form.html.erb
- /nitro_component_transition/app/views/bootstrap/project_tasks/_install_queue_items_table.html.erb
- /nitro_survey/app/views/nitro_survey/blasts/_prompt.html.erb
- /recruiting/app/views/recruiting/candidate_applications/index.html.erb
- /recruiting/app/views/recruiting/retention_report/index.html.erb
- /sales/app/views/sales/reps/qualifiers.html.erb
- /sales/app/views/sales/test_pages/mentor_group_bonuses.html.erb
- /user_profile/app/views/user_profile/user/products.html.erb
- /user_profile/app/views/user_profile/user/sources.html.erb

**React (21 locations, 17 files):**
- /accounting/app/components/invoices/ListTable.js
- /employee_health_check/app/components/HealthCheck/CheckboxInput.jsx
- /employee_health_check/app/components/HealthCheckOptions/OptionsForm.jsx
- /human_resources/app/components/HistoryFlyout/FlagNotes.jsx
- /inventory/app/components/InventoryLocationForm/ProductSelection.js
- /inventory/app/components/ReorderRuleListing/Row.js
- /inventory/app/components/Search/CheckFilter.js
- /inventory/app/containers/InventoryLocationForm.js
- /inventory/app/containers/GenerateBarcodeForm/container.js
- /nitro_react/app/components/CheckboxArrayFormField.jsx
- /nitro_react/app/components/CheckboxFormField.jsx
- /nitro_survey/app/components/SurveyTaker/WrappedInputs.jsx
- /recruiting/app/components/CheckboxInput.jsx
- /support_tickets/app/plugins/project/ProjectFormFields/ProductCheckbox.jsx
- /support_tickets/app/plugins/project/Service/ServiceForm.js
- /user_profile/app/components/MyProfile/CheckboxInput.jsx
- /warranty/app/components/WarrantyServiceOrder/ProjectItemFields.jsx

**Comment by jasoncypret** [comment]
@coleerikson Thanks! As I understand it only the React sections are under intense scrutiny as they are switching to use state. Is that correct?

**Comment by coleerikson** [comment]
@jasoncypret That is correct

**Comment by jasoncypret** [comment]
> * [ ]  /support_tickets/app/plugins/project/ProjectFormFields/ProductCheckbox.jsx - **IDK how to access this..**

image

You can view a support ticket. Its the checkbox for private

**Comment by jasoncypret** [comment]
> * [ ]  /support_tickets/app/plugins/project/ProjectFormFields/ProductCheckbox.jsx - **IDK how to access this..**

image

You can find a home and click the add service ticket button. Try this house in QA: https://nitroqa.powerhrg.com/homes/8047393?current_tab=Projects

**Comment by coleerikson** [comment]
New PR here: https://github.com/powerhome/playbook/pull/1372

**Pull request closed without merging**